Gems/Terrain/Assets/Shaders/Terrain/TerrainPBR_ForwardPass.azsl

```diff
@@ -76,47 +80,62 @@ ForwardPassOutput TerrainPBR_MainPassPS(VSOutput IN)
 
     float viewDistance = length(ViewSrg::m_worldPosition - surface.position);
     float detailFactor = saturate((viewDistance - TerrainMaterialSrg::m_detailFadeDistance) / max(TerrainMaterialSrg::m_detailFadeLength, EPSILON));
-    float2 detailUv = IN.m_uv * TerrainMaterialSrg::m_detailTextureMultiplier;
+    float2 detailUv = IN.m_worldPosition.xy * TerrainMaterialSrg::m_detailTextureMultiplier;
     
     // ------- Normal -------
     float3 macroNormal = normalize(IN.m_normal);
     
     // ------- Macro Color / Normal -------
     float3 macroColor = TerrainMaterialSrg::m_baseColor.rgb;
 
-    // There's a bug that shows up with an NVidia GTX 1660 Super card happening on driver versions as recent as 496.49 (10/26/21) in which
-    // the IN.m_uv values will intermittently "flicker" to 0.0 after entering and exiting game mode.  
-    // (See https://github.com/o3de/o3de/issues/5014)
-    // This bug has only shown up on PCs when using the DX12 RHI.  It doesn't show up with Vulkan or when capturing frames with PIX or
-    // RenderDoc.  Our best guess is that it is a driver bug.  The workaround is to use the IN.m_uv values in a calculation prior to the 
-    // point that we actually use them for macroUv below.  The "if(any(!isnan(IN.m_uv)))" seems to be sufficient for the workaround.  The
-    // if statement will always be true, but just the act of reading these values in the if statement makes the values stable.  Removing
-    // the if statement causes the flickering to occur using the steps documented in the bug.
-    if (any(!isnan(IN.m_uv)))
+    uint2 macroGridResolution = uint2(TerrainSrg::m_macroMaterialGrid.m_resolution >> 16, TerrainSrg::m_macroMaterialGrid.m_resolution & 0xFFFF);
+    float macroTileSize = TerrainSrg::m_macroMaterialGrid.m_tileSize;
+    float2 macroGridOffset = TerrainSrg::m_macroMaterialGrid.m_offset;
+    uint2 macroGridPosition = (surface.position.xy - macroGridOffset) / macroTileSize;
+
+    uint macroTileIndex = macroGridResolution.x * macroGridPosition.y + macroGridPosition.x;
+    static const uint NumMacroMaterialsPerTile = 4;
+    macroTileIndex *= NumMacroMaterialsPerTile;
+
+    [unroll] for (uint i = 0; i < NumMacroMaterialsPerTile; ++i)
```

Comment:
I would think this would just turn into goto statements after unrolling. Testing this in shader-playground, it looks like it unrolls a loop with a similar structure to this by default even without the [unroll] attribute, likely because the NumMacroMaterialsPerTile is only 4.